subsys/canbus/canopen/CO_driver.c

```diff
@@ -0,0 +1,491 @@
+		return;
+	}
+
+	for (i = 0U; i < CANmodule->rx_size; i++) {
+		if (CANmodule->rx_array[i].filter_id != CAN_NO_FREE_FILTER) {
+			can_detach(CANmodule->dev,
+				   CANmodule->rx_array[i].filter_id);
+			CANmodule->rx_array[i].filter_id = CAN_NO_FREE_FILTER;
+		}
+	}
+}
+
+static void canopen_rx_isr_callback(struct zcan_frame *msg, void *arg)
+{
+	CO_CANrx_t *buffer = (CO_CANrx_t *)arg;
+	CO_CANrxMsg_t rxMsg;
+
+	if (!buffer || !buffer->pFunct) {
+		LOG_ERR("failed to process CAN rx isr callback");
+		return;
+	}
+
+	rxMsg.DLC = msg->dlc;
+	memcpy(rxMsg.data, msg->data, msg->dlc);
+	buffer->pFunct(buffer->object, &rxMsg);
+}
+
+static void canopen_tx_isr_callback(u32_t error_flags, void *arg)
+{
+	CO_CANmodule_t *CANmodule = arg;
+
+	if (!CANmodule) {
+		LOG_ERR("failed to process CAN tx isr callback");
+		return;
+	}
+
+	if (error_flags == CAN_TX_OK) {
+		CANmodule->first_tx_msg = false;
+	}
+
+	k_work_submit_to_queue(&canopen_tx_workq, &canopen_tx_queue.work);
+}
+
+static void canopen_tx_retry(struct k_work *item)
+{
+	struct canopen_tx_work_container *container =
+		CONTAINER_OF(item, struct canopen_tx_work_container, work);
+	CO_CANmodule_t *CANmodule = container->CANmodule;
+	struct zcan_frame msg;
+	CO_CANtx_t *buffer;
+	int err;
+	u16_t i;
+
+	CO_LOCK_CAN_SEND();
+
+	for (i = 0; i < CANmodule->tx_size; i++) {
+		buffer = &CANmodule->tx_array[i];
+		if (buffer->bufferFull) {
+			msg.id_type = CAN_STANDARD_IDENTIFIER;
+			msg.std_id = buffer->ident;
+			msg.dlc = buffer->DLC;
+			msg.rtr = (buffer->rtr ? 1 : 0);
+			memcpy(msg.data, buffer->data, buffer->DLC);
+
+			err = can_send(CANmodule->dev, &msg, K_NO_WAIT,
+				       canopen_tx_isr_callback, CANmodule);
+			if (err == CAN_TIMEOUT) {
+				break;
+			} else if (err != CAN_TX_OK) {
+				LOG_ERR("failed to send CAN frame (err %d)",
+					err);
+				CO_errorReport(CANmodule->em,
+					       CO_EM_GENERIC_SOFTWARE_ERROR,
+					       CO_EMC_COMMUNICATION, 0);
+
+			}
+
+			buffer->bufferFull = false;
+		}
+	}
+
+	CO_UNLOCK_CAN_SEND();
+}
+
+void CO_CANsetConfigurationMode(s32_t CANbaseAddress)
+{
+	/* No operation */
+}
+
+void CO_CANsetNormalMode(CO_CANmodule_t *CANmodule)
+{
+	CANmodule->CANnormal = true;
+}
+
+CO_ReturnError_t CO_CANmodule_init(CO_CANmodule_t *CANmodule,
+				   s32_t CANbaseAddress,
+				   CO_CANrx_t rxArray[], u16_t rxSize,
+				   CO_CANtx_t txArray[], u16_t txSize,
+				   u16_t CANbitRate)
+{
+	struct device *device;
+	u16_t i;
+	int err;
+
+	LOG_DBG("rxSize = %d, txSize = %d", rxSize, txSize);
+
+	device = canopen_get_can_device(CANbaseAddress);
+
+	if (!CANmodule || !rxArray || !txArray || !device) {
+		LOG_ERR("failed to initialize CAN module %d", CANbaseAddress);
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	if (rxSize > CONFIG_CAN_MAX_FILTER) {
+		LOG_ERR("insufficient number of concurrent CAN RX filters"
+			" (needs %d, %d available)", rxSize,
+			CONFIG_CAN_MAX_FILTER);
+		return CO_ERROR_OUT_OF_MEMORY;
+	} else if (rxSize < CONFIG_CAN_MAX_FILTER) {
```

Comment:
Are all filters standard frames without mask? If not, the index may increase by more than one and you need more filters. Depending on the filter bank configurations.

Reply:
No, some of the filters are with mask, some without. The number of RX filters needed is determined by the features enabled in the object dictionary (e.g. number of RPDOs) for the given application.

Reply:
I would not do this check. For STM32 you probably need more filters than messages. (because some are with, and some are without a mask.

Reply:
I would like to keep it in as it makes it much easier to adjust the filter allocation for all other CAN controller drivers.
I have converted the output to `LOG_DBG()` instead.